### PR TITLE
15.4 Receive Ring Buffer

### PR DESCRIPTION
**Pull request by tyler-potyondy**
#### Current Design / Issue
The 15.4 driver receive provides a buffer for one packet. Upon receipt of a packet, an upcall is scheduled. The kernel maintains possession of this buffer until the upcall is handled by the userprocess. As such, if packets are received in rapid succession (and the userprocess does not handle the upcall between packets), the second packet will overwrite the first. 

#### Proposed Solution
This PR adds support for the userprocess to pass a ring buffer to the 15.4 driver capsule. The userprocess determines the size of the buffer and the capsule begins filling the buffer. I will not go into the specifics of the design here as I attempted to detail this in depth at the top of the `driver.rs` file. 

With this ring buffer, the capsule possess a mechanism for handling the receipt of multiple packets in quick succession (or with a userprocess being slow to yield and handle upcalls). This design allows flexibility as the userprocess can stipulate the necessary size of the ring buffer.

#### Minor Changes
As part of the ongoing openthread port, the MIC len is useful / needed information. This minor change to the frame metadata passed to the userprocess was lumped in with this PR.

### Testing Strategy

This pull request was tested by successfully joining and remaining attached to an OpenThread network. The libtock-c 15.4 tests will need to be updated once this is upstreamed.

#### Design Choices Worth Further Discussion
I chose to have the ring buffer overwrite the oldest data upon the buffer being depleted. I believe this to be reasonable as the newest data is likely the most pertinent. This also seemed preferable to the alternative of dropping packets. If this assumption is faulty or others disagree, this can be altered.

#### TODO
In a separate PR, I am hoping to add functionality for compilation specified debug logging. As a place holder, I am leaving two helpful debug statements commented out. 

### Documentation Updated

I updated and added comments on this design within the file, but the 15.4 networking documentation will be in need of a revamp in light of the many changes. I believe the 15.4 stack will continue to see additional changes in the coming months so it likely makes sense to hold off on updating until we reach a stable point. 

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1137]: This is concerning. While it's maybe ok to store metadata in a userspace buffer, and tell userspace to not mess with it, using the `[]` operator on slices means it's trivial for userspace to crash the kernel by changing this byte.

One fix is to not use `[]` and instead use get/set. A less rusty version would be to check the indices.

**Review by bradjc — CHANGES_REQUESTED** [review]
It seems like this would require userspace to break the semantics of readwrite allow: userspace would have to read the received packet while the buffer remains "allowed" and the kernel is writing it, when only one of those should happen at a time.

**Comment by tyler-potyondy** [comment]
> It seems like this would require userspace to break the semantics of readwrite allow: userspace would have to read the received packet while the buffer remains "allowed" and the kernel is writing it, when only one of those should happen at a time.

I do not think this breaks the semantics of readwrite allow (although I could be mistaken). The way I attempted to implement this should follow:

process allows buffer -> kernel receives, schedules an upcall, and keeps writing to buffer until unallowed -> upcall is handled, but before reading data the userprocess unallows the buffer

Once userspace has completed reading all active data, it then allows the original buffer again to the kernel. At any given point in time, the userprocess reading and the kernel writing _should_ be mutually exclusive.

 To my knowledge, this is in compliance with the readwrite allow semantics. I agree that this is somewhat cumbersome from the process perspective, but I think it should be possible without breaking the semantics. @bradjc please let me know if I am overlooking some other detail

File: capsules/extra/src/ieee802154/driver.rs
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1137]: Good catch, I overlooked this. I'll go ahead and add this.

**Comment by bradjc** [comment]
> I do not think this breaks the semantics of readwrite allow (although I could be mistaken). The way I attempted to implement this should follow:
> 
> process allows buffer -> kernel receives, schedules an upcall, and keeps writing to buffer until unallowed -> upcall is handled, but before reading data the userprocess unallows the buffer
> 
> Once userspace has completed reading all active data, it then allows the original buffer again to the kernel. At any given point in time, the userprocess reading and the kernel writing _should_ be mutually exclusive.
> 
> To my knowledge, this is in compliance with the readwrite allow semantics. I agree that this is somewhat cumbersome from the process perspective, but I think it should be possible without breaking the semantics. @bradjc please let me know if I am overlooking some other detail

You are correct, this is fine. But what happens if a packet arrives while the app is handling an older packet? ie

process allows buffer -> kernel receives, schedules an upcall, and keeps writing to buffer until unallowed -> upcall is handled, but before reading data the userprocess unallows the buffer -> while app handles, new packet arrives -> kernel does what?

**Comment by tyler-potyondy** [comment]
> You are correct, this is fine. But what happens if a packet arrives while the app is handling an older packet? 

In this case, the kernel will drop any packets received and not modify the buffer due to this guard statement:

```
if rbuf.len() == 0 || (rbuf.len() - RING_BUF_METADATA_SIZE) % USER_FRAME_MAX_SIZE != 0 { return false }
```

rbuf.len() will be zero once the process unallows the buffer

**Comment by bradjc** [comment]
> the kernel will drop any packets received

Why not setup two buffers so this doesn't happen? It seems like the ring buffer is supposed to help prevent dropped packets, so I'm a little confused why not avoid this case.

**Comment by tyler-potyondy** [comment]
> Why not setup two buffers so this doesn't happen? It seems like the ring buffer is supposed to help prevent dropped packets, so I'm a little confused why not avoid this case.

This is entirely possible with this design and is something the userprocess could configure. The ring buffer's primary purpose was to solve the problem of packets being overwritten before the upcall was handled. 

To prevent dropped packets while handling the upcall, the userprocess would do the following:

1. Create ring buffer and allow to kernel
2. Kernel receives packets, placing them in the ring buffer, and schedules upcall on first packet receipt
3. Userprocess handles upcall, first unallowing the buffer and notifying the kernel the upcall has been handled with command syscall (this here can be altered to allow a 2nd ring buffer used while handling the primary ring buffer)

Your question @bradjc gets at a bigger question with 15.4 receiving that places the primary burden on the userprocess to provide correct buffers at the correct time. After each received packet, the userprocess must call the receive function again which once again begins the process of allowing the buffer.

For discussion's sake here's an outline of the current 15.4 receive handling (w/o the ring buffer):

1. userprocess allows read/write buffer of 129 bytes (2 bytes for metadata and 127 byte 15.4 packet)
2. kernel (on receive) places the packets into the user buffer and then schedules the upcall
3. userprocess handles upcall with provided callback (in the 15.4 example the userprocess unallows the buffer, prints the buffer, and then calls receive again)

**Comment by ppannuto** [comment]
Bigger than this PR, this maybe gets at the need for a more nuanced interface between kernel and userspace; perhaps even a dedicated 'ring buffer allow'. This type of resource management is not going to be unique to networking.

I think the semantics we want might look something like:
 - Userspace `RingAllow`s one large buffer; the whole of the buffer is now owned by the kernel
 - The kernel writes into the buffer, one or more packets [or sensor readings, or audio samples, or...] and adjusts pointers s.t. the portion of the buffer with data is now owned by userspace, while the rest is owned by the kernel still
 - Eventually, an upcall runs and userspace reads the data from the portion of the buffer it currently owns
     - Userspace then makes a syscall to 'return' the read parts of the buffer to the kernel

Overall, I think the data structure is a single ring buffer with pointers _all managed/owned by the kernel_ that define the boundaries of the three possible regions within the buffer: 
```
 head                                  Uptr                                Kptr                                   tail
 |--- owned by userspace w/ data ------ | ---- owned by kernel w/ data ---- | ---- owned by kernel w/out data -----|
```
Where `Kptr` moves whenever the kernel writes data into the buffer, and `Uptr` moves whenever an upcall actually passes data to userspace. `head` moves whenever userspace returns some buffer to the kernel.

---

While a double-buffer (or double-ring-buffer) also works, I feel like it's harder to size buffers well in that case [each buffer has to be large enough for worst-case latency of packet backlog]. A single buffer with moving pointers I think would use the buffer more optimally.

**Comment by tyler-potyondy** [comment]
@ppannuto this sounds like a great idea! I can put together a draft of this design. 

For the purposes of this PR, this would be blocked on completing and upstreaming the ring buffer design. Given the not too distant CPS-IoT OpenThread deadline, is this PR something we should merge in the nearer term? We can later change this receive implementation once a generic ring buffer allow has been merged.

**Comment by bradjc** [comment]
> * Create ring buffer and allow to kernel
> 
> * Kernel receives packets, placing them in the ring buffer, and schedules upcall on first packet receipt
> 
> * Userprocess handles upcall, first unallowing the buffer and notifying the kernel the upcall has been handled with command syscall (this here can be altered to allow a 2nd ring buffer used while handling the primary ring buffer)

Ok great this makes sense. Thanks for spelling that out for me. Given the rate at which 15.4 packets are likely to arrive this seems like a good solution.

I do think we need to revisit how we do allows, although not right now and not in this PR. Let's get something working.

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1047]: It seems like command 28 stops receive in a way. Why not just make this explicitly to declare "the rx buffer is new" so the kernel knows to reset the ring structure?

Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1047]: When the kernel receives this command, the userprocess _should_ have already unallowed the buffer so the kernel has nothing it is able to even reset.

The motivation for having this was to prevent repeated upcalls. Previously, if multiple packets arrived quickly, two upcalls would be scheduled (the ring buffer avoids the upcalls being overwritten but does not guard against this). We do not want this to occur because this would mean the userprocess would be unable to transmit (since the receive function used both the receive and transmit upcall allocation). The `pending_rx` flag stops this from occuring. The side effect of this, however, is we need a mechanism to notify the kernel it can once again schedule an upcall (command 28).

@bradjc does this add some clarity for what this is doing? Would it be more clear to change this flag to `pending_rx_upcall`?

Comment (bradjc) [reply capsules/extra/src/ieee802154/driver.rs:1047]: Why not just unsubscribe the upcall?

**Comment by bradjc** [comment]
This is what I'm imagining from this discussion:

The app:
- Calls allow, then subscribe, then new_buffer command, then RX command.
- On upcall, unallows buffer, calls new_buffer command, and then allows new buffer.
  - This is clumsy to force the correct ordering, where now matter when a packet is received, the packet is either completely dropped (no buffer is allowed) or is correctly added to a ringbuf, but is never added to a new unitiialized ringbuf the kernel thinks is old, or added to an old buffer the kernel thinks is new. Here "old" means the kernel has already been using it as a ringbuf and has stored a packet in it.
- Can unsubscribe upcall if needed.

On a received packet, the kernel:

```rust
app.get_allow_buffer(|abuf| {
  if self.fresh_buffer.get() {
    abuf[0] = 0; // reset ring buffer metadata.
    self.fresh_buffer.set(false);
  }

  // fill packet into ring buffer.

  // signal upcall
  app.upcall(...);
});
```

The motivation being we make explicit (using command) the allow semantics we want, so we know for the future what allow should look like. In this case we want to know if an allow buffer is "new".

File: capsules/extra/src/ieee802154/driver.rs
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1047]: As I understand how unsubscribe works, I do not think this would solve the issue `pending_rx` and command 28 address (2 upcalls being scheduled while receiving). The current doubly scheduled receive upcall is happening when:

1. user process subscribes / allows buffer
2. kernel receives packet 1, places into buffer, and schedules upcall
3. kernel receives packet 2, places into buffer, and schedules another upcall (since the upcall count is two, two upcalls can be scheduled)

Unsubscribing the upcall would need to occur by the userprocess when the upcall is handled (at which time the second upcall has already been scheduled).

**Comment by tyler-potyondy** [comment]
> This is what I'm imagining from this discussion:

This sounds reasonable to me. I'll go ahead and create an issue so we can discuss this further. I think we should spend some time thinking about ways to improve the "clumsiness" you alluded to that enforces the correct ordering of steps.

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [reply capsules/extra/src/ieee802154/driver.rs:1047]: The first argument to `schedule_upcall()` is the upcall index. So the driver having two upcalls means there are two separate upcalls, but each can be scheduled multiple times. The only limit is when the task queue for the process fills up (which right now has 10 spots https://github.com/tock/tock/blob/01918fac039fd6d37c4cea1b3bdbd9ac2683af3f/kernel/src/process_standard.rs#L1337).

I actually think unsubscribing _would_ solve the problem, as when the new subscribe (aka the unsubscribe) happens, we clear any pending upcalls for the old upcall (https://github.com/tock/tock/blob/01918fac039fd6d37c4cea1b3bdbd9ac2683af3f/kernel/src/kernel.rs#L977). Adding the command won't, since the upcall is still pending in the queue waiting for the app to call yield.

Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1047]: I understand now. This is a great idea! I will make the changes to incorporate this instead of the command.

**Comment by tyler-potyondy** [comment]
I encountered an interesting bug today (with the userspace implementation). The upcalls being handled by the userprocess too quickly creates the need for 2 user space ring buffers as @bradjc mentioned earlier. This further highlights the need for a ring buffer allow. An ad hoc implementation of the standard buffer will not work in the long term.

**Comment by tyler-potyondy** [comment]
@bradjc I removed command 28 and incorporated your feedback for clearing the upcalls. With these changes, I faced some timing issues on the libtock-c side of the world. I was able to rework the platform abstraction layer in libtock-c to fix the timing issues and openthread is once again successfully joining and staying attached to a thread network using the most recent version of this branch.

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1143]: As long as `write_index` and `offset` are derived from content in the buffer, there cannot be any `rbuf[]` in this function.

Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1143]: Does confirming `if read_index >= max_pending_rx || write_index >= max_pending_rx` not allow for `rbuf[]`? 

My understanding was that we want to avoid invalid indexing into `rbuf` in the instance when the process provides invalid meta data (read / write index). By checking that the read and write index are valid values, does it not allow us now to index safely into `rbuf`?

(I am assuming that since the capsule is not preempted, this will execute to completion after we verify the parameters and that the userprocess will not have an opportunity to alter the data between when we error check and then perform indexing into `rbuf`. This may however be an inaccurate understanding.)

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1109]: What happens if `rbuf.len()` is 1?

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [reply capsules/extra/src/ieee802154/driver.rs:1143]: That's fine. But, it does mean we have to be very careful since all checking is on us (like with C) and not with the compiler (which is happy to panic).

File: capsules/extra/src/ieee802154/driver.rs
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1109]: That case should fail, but it appears it may slip through since `rbuf.len()` and `RING_BUF_METADATA_SIZE` are both unsigned. I'll update this to account for that case. Good catch!

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:37]:
```suggestion
//! ```text
//! | read index | write index | user_frame 0 | user_frame 1 | ... | user_frame n |
//! ```
```

**Review by bradjc — DISMISSED** [review]

File: capsules/extra/src/ieee802154/driver.rs
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1153]: Should we update `read_index` here? It might cause odd behavior if a new packet is "received" before an old one.

**Review by ppannuto — APPROVED** [review]
+1, this looks like a good interim solution until the kernel can provide something more general and a bit less awkward [i.e., the 'unsubscribe/resubscribe'].

**Comment by alevy** [comment]
This is probably something that the @tock/networking-wg should look at, but give the number of approvals (and I'm adding mine), I think best to merge now.

**Pull request merged**
